### PR TITLE
Bug/fix internal flag in binary sensor

### LINKED ISSUE #1224: Internal flag not properly set for binary sensor
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

Esphome Version: 1.14.3

binary sensor

Cannot change flag internal of binary_sensor, whatever is set element is not visible in Home Assistant

**Problem-relevant YAML-configuration entries:**
```yaml
esphome:
    name: test
    platform: ESP8266
    board: d1_mini_lite
    build_path: target/test

binary_sensor:
  - platform: gpio
    id: bs_1
    name: "test bs1"
    internal: true
    pin:
      number: D0
  - platform: gpio
    id: bs_2
    name: "test bs2"
    internal: false
    pin:
      number: D1

switch:
  - platform: gpio
    id: sw_1
    name: "sw_1"
    internal: true
    pin:
      number: D0
  - platform: gpio
    id: sw_2
    name: "sw_2"
    internal: false
    pin:
      number: D0

```

**Logs (if applicable):**

```c
// Auto generated code by esphome
// ========== AUTO GENERATED INCLUDE BLOCK BEGIN ===========
#include "esphome.h"
using namespace esphome;
using namespace binary_sensor;
using namespace switch_;
gpio::GPIOBinarySensor *bs_1;
gpio::GPIOBinarySensor *bs_2;
gpio::GPIOSwitch *sw_1;
gpio::GPIOSwitch *sw_2;
// ========== AUTO GENERATED INCLUDE BLOCK END ==========="

void setup() {
  // ===== DO NOT EDIT ANYTHING BELOW THIS LINE =====
  // ========== AUTO GENERATED CODE BEGIN ===========
  // esphome:
  //   name: test
  //   platform: ESP8266
  //   board: d1_mini_lite
  //   build_path: target/test
  //   libraries: []
  //   board_flash_mode: dout
  //   arduino_version: espressif8266@2.2.3
  //   includes: []
  //   platformio_options: {}
  //   esp8266_restore_from_flash: false
  App.pre_setup("test", __DATE__ ", " __TIME__);
  // binary_sensor:
  // switch:
  // binary_sensor.gpio:
  //   platform: gpio
  //   id: bs_1
  //   name: test bs1
  //   internal: true
  //   pin:
  //     number: 16
  //     inverted: false
  //     mode: INPUT
  bs_1 = new gpio::GPIOBinarySensor();
  App.register_component(bs_1);
  // binary_sensor.gpio:
  //   platform: gpio
  //   id: bs_2
  //   name: test bs2
  //   internal: false
  //   pin:
  //     number: 5
  //     inverted: false
  //     mode: INPUT
  bs_2 = new gpio::GPIOBinarySensor();
  App.register_component(bs_2);
  // switch.gpio:
  //   platform: gpio
  //   id: sw_1
  //   name: sw_1
  //   internal: true
  //   pin:
  //     number: 16
  //     inverted: false
  //     mode: OUTPUT
  //   interlock_wait_time: 0ms
  //   restore_mode: RESTORE_DEFAULT_OFF
  sw_1 = new gpio::GPIOSwitch();
  App.register_component(sw_1);
  // switch.gpio:
  //   platform: gpio
  //   id: sw_2
  //   name: sw_2
  //   internal: false
  //   pin:
  //     number: 16
  //     inverted: false
  //     mode: OUTPUT
  //   interlock_wait_time: 0ms
  //   restore_mode: RESTORE_DEFAULT_OFF
  sw_2 = new gpio::GPIOSwitch();
  App.register_component(sw_2);
  App.register_binary_sensor(bs_1);
  bs_1->set_name("test bs1");
  bs_1->set_internal("internal");
  App.register_binary_sensor(bs_2);
  bs_2->set_name("test bs2");
  bs_2->set_internal("internal");
  App.register_switch(sw_1);
  sw_1->set_name("sw_1");
  sw_1->set_internal(true);
  App.register_switch(sw_2);
  sw_2->set_name("sw_2");
  sw_2->set_internal(false);
  bs_1->set_pin(new GPIOPin(16, INPUT, false));
  bs_2->set_pin(new GPIOPin(5, INPUT, false));
  sw_1->set_pin(new GPIOPin(16, OUTPUT, false));
  sw_1->set_restore_mode(gpio::GPIO_SWITCH_RESTORE_DEFAULT_OFF);
  sw_2->set_pin(new GPIOPin(16, OUTPUT, false));
  sw_2->set_restore_mode(gpio::GPIO_SWITCH_RESTORE_DEFAULT_OFF);
  // =========== AUTO GENERATED CODE END ============
  // ========= YOU CAN EDIT AFTER THIS LINE =========
  App.setup();
}

void loop() {
  App.loop();
}

```

**Additional information and things you've tried:**
Please look at set_internal call
```c
  bs_1->set_internal("internal");
  bs_2->set_internal("internal");
  sw_1->set_internal(true);
  sw_2->set_internal(false);
```

### PR DESCRIPTION
**Pull request by pkuehne**
## Description:

The `internal` flag on binary sensors isn't set correctly. This is because the _name_ of the flag is used when generating the code instead of the _value_. This works correctly now.

Additionally, I took the liberty of creating a test to ensure that this will continue to work. The test reads a config file, let's the `core` component do all the assembly of the C++ code and then validates that the expected statements are in that code. To do this without having to write out all the C++ code to disk, I had to split the `write_cpp` function in `__main__.py` into two. One to assemble the C++ statements inside `CORE` and the other to write them to disk.

The test setup is very simple and could even be abstracted into a fixture like this:
```python
def test_setup_function(generate_main):
    maincpp = generate_main("test.yaml")

    assert "void setup() {" in maincpp
```

This would then ensure that the only code that is actually generated can be properly tested. I can do that in this PR if that would be useful.

**Related issue (if applicable):** fixes esphome/issues#1224

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

### CODE REVIEW

File: tests/component_tests/binary_sensor/test_binary_sensor.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    CORE.config_path = "tests/component_tests/binary_sensor/test_binary_sensor.yaml"
+    CORE.config = read_config({})
+
+    # When
+    assemble_cpp_file(CORE.config)
+    # print(CORE.cpp_main_section)
+
+    # Then
+    assert "bs_1->set_internal(true);" in CORE.cpp_main_section
+    assert "bs_2->set_internal(false);" in CORE.cpp_main_section
```
Comment (OttoWinter) [inline tests/component_tests/binary_sensor/test_binary_sensor.py:22]: This should call `CORE.reset()` at the end.

File: esphome/__main__.py [hunk 2]
```diff
@@ -131,6 +131,11 @@ def wrapped(conf):
 
 
 def write_cpp(config):
+    assemble_cpp_file(config)
+    write_cpp_file()
+
+
+def assemble_cpp_file(config):
```
Comment (OttoWinter) [inline esphome/__main__.py:138]: Assemble is a bit ambiguous here (other use: assembler as in compiling assembler code)

Better would be `generate_cpp_contents()` or something like that.

File: tests/component_tests/binary_sensor/test_binary_sensor.py (hunk 1 above)
Comment (pkuehne) [reply tests/component_tests/binary_sensor/test_binary_sensor.py:22]: ah yes, forgot this was a global

File: esphome/__main__.py (hunk 2 above)
Comment (pkuehne) [reply esphome/__main__.py:138]: will do!

File: esphome/components/binary_sensor/__init__.py [hunk 3]
```diff
@@ -224,7 +224,7 @@ def validate_multi_click_timing(value):
 def setup_binary_sensor_core_(var, config):
     cg.add(var.set_name(config[CONF_NAME]))
     if CONF_INTERNAL in config:
-        cg.add(var.set_internal(CONF_INTERNAL))
+        cg.add(var.set_internal(config[CONF_INTERNAL]))
```
Comment (OttoWinter) [inline esphome/components/binary_sensor/__init__.py:227]: Thanks for spotting this btw! Nice find!

**Review by OttoWinter — APPROVED** [review]

File: esphome/components/binary_sensor/__init__.py (hunk 3 above)
Comment (pkuehne) [reply esphome/components/binary_sensor/__init__.py:227]: I can't take credit for that, it was reported in esphome/issues#1224 with a literal patch. I just added the unit tests on top

**Pull request merged**